README.md
# tv7SamsungNg

This Samsung smart TV app is implemented using __Angular 11__. Project was generated with the [Angular CLI](https://cli.angular.io/) version __11.2.1__. Application is tested with the following devices:
  - Samsung smart TV (version 5.0)
  - Samsung TV emulator (version 3.0)
  - Samsung TV emulator (version 4.0)
  - Samsung TV emulator (version 6.0)

## Similar app on the other TV platforms
  - [LG](https://github.com/heaven-dev/tv7LgSpa)
  - Android:
    - [fi](https://github.com/heaven-dev/taivasTv7Android)
    - [et](https://github.com/heaven-dev/taevasTv7Android)
    - [ru](https://github.com/heaven-dev/nebesaTv7Android)
    - [sv](https://github.com/heaven-dev/himlenTv7Android)

## Demo

App running on Samsung TV emulator (Tizen version 6.0).

![Demo](https://github.com/heaven-dev/demo/blob/main/demo/demo.gif)

## Overview

__TV7__ web sites:
  - [Taivas TV7](https://www.tv7.fi/)
  - [Taevas TV7](https://www.tv7.ee/)
  - [Himlen TV7](https://www.himlentv7.se/)
  - [Небеса ТВ7](https://www.nebesatv7.com/)

Application make possible to watch TV channel and videos from the video archive. The application supports the following locales:
  - __fi__: Finnish
  - __et__: Estonian
  - __sv__: Swedish
  - __ru__: Russian

Each locale is an independent application. You have to use __gulp__ to generate a localized version of the application. Instructions are below.

## Instructions

### Download and install git
  - If your computer OS is windows you can download the git from [here](https://git-scm.com/download/win).
  - If your computer OS is linux (Ubuntu) you can install git using package manager.

### Clone repository
Clone this repository to your computer disk.
  - __git clone https://github.com/heaven-dev/tv7SamsungNg.git__

### Download and install node (version 12.x.x)
  - If your computer OS is windows you can download node from [here](https://nodejs.org/en/download/).
  - If your computer OS is linux (Ubuntu) you can install node using package manager.

### Install dependencies
  - In the root folder of the project: __npm install__

### Install gulp-cli
  - __sudo npm install -g gulp-cli__
  - Gulp is used to build the application:
    - Build __finnish__: __npm run build-fi__
    - Build __estionian__: __npm run build-et__
    - Build __swedish__: __npm run build-sv__
    - Build __russian__: __npm run build-ru__
    - Build __all__: __npm run build-all__
  - Build is created under the __root folder__ of this project for example __tizen/dist/fi__ and respectively to each locale.
      
### Download and install Tizen studio
  - Download Tizen studio from [here](https://developer.tizen.org/development/tizen-studio/download).
    - Select your OS.
    - Download this package: __Tizen Studio 4.1 with IDE installer__.
    - Save the package to the disk and start the setup.
    - Follow the installation instructions from [here](https://developer.samsung.com/smarttv/develop/getting-started/setting-up-sdk/installing-tv-sdk.html).
    - When you see text __To install the extensions from local images:__ you have done.
  - Alternatively you can also use [Atom](https://developer.samsung.com/smarttv/develop/extension-libraries/atom-ide-package.html) or [vscode](https://developer.samsung.com/smarttv/develop/extension-libraries/vscode-extension.html).

### Import projects into the Tizen studio IDE
  - Start Tizen studio.
  - Select workspace folder and launch.
  - Select __File -> Import -> Tizen project__.
    - Select location for example (__projectRoot/tizen/dist/fi__). Similarly you can import each project to select folder of the locale for example __projectRoot/tizen/dist/et__, __projectRoot/tizen/dist/sv__ or __projectRoot/tizen/dist/ru__.
    - Select __Next__.
    - Select profile __samsung-tv__ and version __6.0__ and select __Next__.
  - Now you should see the project on IDE.
  - All __HTML__, __CSS__ and __javascript__ files are compressed by the build commands.

### Run application on emulator
  - Select __Run -> Run configurations__ from the menu.
  - Select Tizen web application and click __New launch configuration__.
  - Set name to launch configuration.
  - Select project if not already selected.
  - Select __Apply__ and __Close__.
  - You can run __release__ version:
    - Click the project name with right mouse button.
    - Select __Run As -> Run configurations__ and select launch configuration you create and click __Run__.
    - Emulator is started and application is opened into the emulator.
  - You can run __debug__ version:
    - Click the project name with right mouse button.
    - Select __Debug As -> Debug configurations__ and select launch configuration you create and click __Run__.
    - Debug mode opens also an inspector of the browser.
    - Emulator is started and application is opened into the emulator.

### Run application on TV
  - At first you have to switch TV to the developer mode.
  - Your computer and TV have to be in the same network (wi-fi is fine).
  - Follow instructions from [here](https://developer.samsung.com/smarttv/develop/getting-started/using-sdk/tv-device.html).
  - Be sure that your computer firewall doesn't block the connection between the computer and the TV.
  - Probably you have to a create certificate profile (__Tools -> Certificate Manager__) and a Samsung account to run the application on TV.

### Run application on a browser
  - Prepare app by editing this line: [locale.service.ts](https://github.com/heaven-dev/tv7SamsungNg/blob/main/src/app/services/locale.service.ts#L7)
    - Add one of the following locale (__fi__, __et__, __ru__ or __sv__) to the file.
    - You don't need to care about this file if you build the app with a build command for example __npm run build-fi__. In that case the gulp automatically edits the file.
    - Set [this](https://github.com/heaven-dev/tv7SamsungNg/blob/main/src/app/helpers/constants.ts#L1) flag to __true__.
  - Run command __ng serve__ in the project root.
  - Navigate to __http://localhost:4200__ and the app is starting.

### TV or emulator Tizen version required to run this application 
  - __Tizen version 3.0 TV or newer__ (release year of Tizen version 3.0 TV is 2017)
    - Tizen TV [web engine](https://developer.samsung.com/smarttv/develop/specifications/web-engine-specifications.html)
    - Tizen TV [general specifications](https://developer.samsung.com/smarttv/develop/specifications/general-specifications.html)

### Useful links
  - [Samsung Smart TV](https://developer.samsung.com/smarttv/develop/getting-started/quick-start-guide.html)
  - [Tizen TV](https://developer.tizen.org/tizen/tv) developer web pages.
  - [Command Line Interface (CLI) Commands](https://developer.tizen.org/development/tizen-studio/web-tools/cli)
  - [Application Signing and Certificates](https://docs.tizen.org/application/web/tutorials/sign-certificate/)
  - [Tizen TV Web Device API Reference](https://docs.tizen.org/application/web/api/latest/device_api/tv/index.html)

### License

CHANGELOG.md
# 20.07.2021 15:45

### Bug Fixes

* none

### Features

* Added dynamic category rows to archive main page.
* New version number (2.3.0).


# 13.07.2021 11:50

### Bug Fixes

* none

### Features

* Configuration clear functionality added to the information page.
* New version number (2.2.9).


# 10.07.2021 14:16

### Bug Fixes

* none

### Features

* Added series programs to the archive main page.
* Added series page which can be used to add/remove series to favorites.
* New version number (2.2.8).


# 05.07.2021 11:34

### Bug Fixes

* none

### Features

* Search string appended to the title of search result page.
* Search history is available on the search page.
* New version number (2.2.7).


# 01.07.2021 12:40

### Bug Fixes

* none

### Features

* Font sizes calculated on the platform info page.
* New version number (2.2.6).


# 17.05.2021 09:42

### Bug Fixes

* Fix to the startOtherVideo() method of the archive player.

### Features

* New version number (2.2.5).


# 29.04.2021 08:14

### Bug Fixes

* Fixed channel info page scroll bug.

### Features

* New version number (2.2.4).


# 26.04.2021 18:56

### Bug Fixes

* Fixed tv player and archive player 4:3 aspect ratio video bug.

### Features

* New version number (2.2.3).


# 24.04.2021 16:12

### Bug Fixes

* Fixed platform info page keydown event remove bug.

### Features

* New version number (2.2.2).


# 21.04.2021 09:39

### Bug Fixes

* Fixed pause/play bug on archive player page.

### Features

* New version number (2.2.1).


# 03.04.2021 16:46

### Bug Fixes

* Fix to archive player component.

### Features

* New version number (2.2.0).


# 03.04.2021 16:46

### Bug Fixes

* Fix to archive player component.

### Features

* New version number (2.1.15).


# 03.04.2021 14:36

### Bug Fixes

* Fix to archive player component.

### Features

* New version number (2.1.14).


# 01.04.2021 18:34

### Bug Fixes

* Fallback image changed.

### Features

* New version number (2.1.13).


# 01.04.2021 17:44

### Bug Fixes

* Fix to styles of tv main page.

### Features

* New version number (2.1.12).


# 01.04.2021 12:14

### Bug Fixes

* none

### Features

* Changed sidebar to use absolute positioning.
* New version number.


# 01.04.2021 08:14

### Bug Fixes

* none

### Features

* Added play icon to focused other video item on archive player page.
* New version number.


# 31.03.2021 17:58

### Bug Fixes

* none

### Features

* Added own handler to rewind and forward events on archive player.
* New version number.


# 31.03.2021 11:39

### Bug Fixes

* none

### Features

* Added possibility to open another video on archive player during playback.
* New version number.


# 30.03.2021 12:02

### Bug Fixes

* none

### Features

* Added media error handler to archive and tv player.
* New version number.


# 29.03.2021 14:32

### Bug Fixes

* none

### Features

* Added restart button to error page.
* New version number.


# 26.03.2021 09:10

### Bug Fixes

* none

### Features

* Changes to archive and tv player components.
* New version number.


# 25.03.2021 12:03

### Bug Fixes

* none

### Features

* Changes to archive and tv player components.
* New version number.


# 23.03.2021 08:03

### Bug Fixes

* none

### Features

* Added videojs event handlers. Change to error handling.
* Changed STOP remote key functionality.
* Reconnecting component renamed.
* Change to keydown handler and styles of error page.
* Changes to stream reconnecting functionality.

src/app/services/locale.service.ts
import { Injectable } from '@angular/core';
import { CommonService } from './common.service';

/**
 * Updated by gulp script during a build of the app.
 */
const selectedLocale: string = 'fi';

/**
 * Updated by gulp script during a build of the app.
 */
const appName: string = 'Taivas TV7';

/**
 * Updated by gulp script during a build of the app.
 */
const appVersion: string = '2.3.0';

export const localeFi: string = 'fi';
export const localeEt: string = 'et';
export const localeRu: string = 'ru';
export const localeSv: string = 'sv';

const logoTaivas: string = 'assets/logo_taivas.png';
const logoTaevas: string = 'assets/logo_taevas.png';
const logoNebesa: string = 'assets/logo_nebesa.png';
const logoHimlen: string = 'assets/logo_himlen.png';

const archivePageImageTaivas: string = 'assets/archive_page_image_taivas.png';
const archivePageImageTaevas: string = 'assets/archive_page_image_taevas.png';
const archivePageImageNebesa: string = 'assets/archive_page_image_nebesa.png';
const archivePageImageHimlen: string = 'assets/archive_page_image_himlen.png';

const channelUrlFi: string = 'https://vod.tv7.fi:443/tv7-fi/_definst_/smil:tv7-fi.smil/playlist.m3u8';
const channelUrlEt: string = 'https://vod.tv7.fi:443/tv7-ee/_definst_/smil:tv7-ee.smil/playlist.m3u8';
const channelUrlRu: string = 'https://vod.tv7.fi:443/tv7-ru/_definst_/smil:tv7-ru.smil/playlist.m3u8';
const channelUrlSv: string = 'https://vod.tv7.fi:443/tv7-se/_definst_/smil:tv7-se.smil/playlist.m3u8';

const archiveUrlFi: string = 'https://www.tv7.fi/wp-json/tv7-api/v1/';
const archiveUrlEt: string = 'https://www.tv7.ee/wp-json/tv7-api/v1/';
const archiveUrlRu: string = 'https://www.nebesatv7.com/wp-json/tv7-api/v1/';
const archiveUrlSv: string = 'https://www.himlentv7.se/wp-json/tv7-api/v1/';

const archiveLanguageFi: string = 'FI1';
const archiveLanguageEt: string = 'ET1';
const archiveLanguageRu: string = 'RU1';
const archiveLanguageSv: string = 'SV1';

const archivePageCategoriesFi: any = [
  11,
  6,
  15,
  18,
  17,
  8,
  12,
  5,
  25,
  1,
  7,
  2,
  13,
  4,
  10,
  9,
  14,
  3
];
const archivePageCategoriesEt: any = [
  24,
  8,
  13,
  4,
  1,
  7,
  2,
  6,
  3,
  15,
  5
];
const archivePageCategoriesRu: any = [
  26,
  1,
  6,
  3,
  10,
  35,
  25,
  8,
  15,
  4,
  14
];
const archivePageCategoriesSv: any = [
  1,
  10,
  9,
  6,
  3,
  8,
  15,
  7,
  11,
  5,
  25,
  18,
  4,
  23,
  2
];

const keyboardLetters: any = {
  1: [
    { n: 'q', c: 'Q' },
    { n: 'w', c: 'W' },
    { n: 'e', c: 'E' },
    { n: 'r', c: 'R' },
    { n: 't', c: 'T' },
    { n: 'y', c: 'Y' },
    { n: 'u', c: 'U' },
    { n: 'i', c: 'I' },
    { n: 'o', c: 'O' },
    { n: 'p', c: 'P' },
    { n: 'å', c: 'Å' }
  ],
  2: [
    { n: 'a', c: 'A' },
    { n: 's', c: 'S' },
    { n: 'd', c: 'D' },
    { n: 'f', c: 'F' },
    { n: 'g', c: 'G' },
    { n: 'h', c: 'H' },
    { n: 'j', c: 'J' },
    { n: 'k', c: 'K' },
    { n: 'l', c: 'L' },
    { n: 'ö', c: 'Ö' },
    { n: 'ä', c: 'Ä' }
  ],
  3: [
    { n: 'z', c: 'Z' },
    { n: 'x', c: 'X' },
    { n: 'c', c: 'C' },
    { n: 'v', c: 'V' },
    { n: 'b', c: 'B' },
    { n: 'n', c: 'N' },
    { n: 'm', c: 'M' }
  ],
};

const keyboardNumberSpecial: any = {
  1: [
    { n: '1' },
    { n: '2' },
    { n: '3' },
    { n: '4' },
    { n: '5' },
    { n: '6' },
    { n: '7' },
    { n: '8' },
    { n: '9' },
    { n: '0' },
    { n: ',' },
  ],
  2: [
    { n: '.' },
    { n: ';' },
    { n: ':' },
    { n: '!' },
    { n: '=' },
    { n: '/' },
    { n: '(' },
    { n: ')' },
    { n: '[' },
    { n: ']' },
    { n: '-' }
  ],
  3: [
    { n: '_' },
    { n: '*' },
    { n: '>' },
    { n: '<' },
    { n: '#' },
    { n: '?' },
    { n: '+' },
    { n: '&' },
  ]
};

const keyboardLettersEt: any = {
  1: [
    { n: 'q', c: 'Q' },
    { n: 'w', c: 'W' },
    { n: 'e', c: 'E' },
    { n: 'r', c: 'R' },
    { n: 't', c: 'T' },
    { n: 'y', c: 'Y' },
    { n: 'u', c: 'U' },
    { n: 'i', c: 'I' },
    { n: 'o', c: 'O' },
    { n: 'p', c: 'P' },
    { n: 'ü', c: 'Ü' },
    { n: 'õ', c: 'Õ' }
  ],
  2: [
    { n: 'a', c: 'A' },
    { n: 's', c: 'S' },
    { n: 'd', c: 'D' },
    { n: 'f', c: 'F' },
    { n: 'g', c: 'G' },
    { n: 'h', c: 'H' },
    { n: 'j', c: 'J' },
    { n: 'k', c: 'K' },
    { n: 'l', c: 'L' },
    { n: 'ö', c: 'Ö' },
    { n: 'ä', c: 'Ä' }
  ],
  3: [
    { n: 'z', c: 'Z' },
    { n: 'x', c: 'X' },
    { n: 'c', c: 'C' },
    { n: 'v', c: 'V' },
    { n: 'b', c: 'B' },
    { n: 'n', c: 'N' },
    { n: 'm', c: 'M' }
  ],
};

const keyboardNumberSpecialEt: any = {
  1: [
    { n: '1' },
    { n: '2' },
    { n: '3' },
    { n: '4' },
    { n: '5' },
    { n: '6' },
    { n: '7' },
    { n: '8' },
    { n: '9' },
    { n: '0' },
    { n: ',' },
  ],
  2: [
    { n: '.' },
    { n: ';' },
    { n: ':' },
    { n: '!' },
    { n: '=' },
    { n: '/' },
    { n: '(' },
    { n: ')' },
    { n: '[' },
    { n: ']' },
    { n: '-' }
  ],
  3: [
    { n: '_' },
    { n: '*' },
    { n: '>' },
    { n: '<' },
    { n: '#' },
    { n: '?' },
    { n: '+' },
    { n: '&' },
  ]
};

const keyboardLettersRu: any = {
  1: [
    { n: 'ё', c: 'Ё' },
    { n: 'ъ', c: 'Ъ' },
    { n: 'я', c: 'Я' },
    { n: 'ш', c: 'Ш' },
    { n: 'е', c: 'Е' },
    { n: 'р', c: 'Р' },
    { n: 'т', c: 'Т' },
    { n: 'ы', c: 'Ы' },
    { n: 'у', c: 'У' },
    { n: 'и', c: 'И' },
    { n: 'о', c: 'О' },
    { n: 'п', c: 'П' },
    { n: 'ю', c: 'Ю' },
  ],
  2: [
    { n: 'щ', c: 'Щ' },
    { n: 'э', c: 'Э' },
    { n: 'а', c: 'А' },
    { n: 'с', c: 'С' },
    { n: 'д', c: 'Д' },
    { n: 'ф', c: 'Ф' },
    { n: 'г', c: 'Г' },
    { n: 'ч', c: 'Ч' },
    { n: 'й', c: 'Й' },
    { n: 'к', c: 'К' },
    { n: 'л', c: 'Л' },
    { n: 'ь', c: 'Ь' }
  ],
  3: [
    { n: 'ж', c: 'Ж' },
    { n: 'з', c: 'З' },
    { n: 'х', c: 'Х' },
    { n: 'ц', c: 'Ц' },
    { n: 'в', c: 'В' },
    { n: 'б', c: 'Б' },
    { n: 'н', c: 'Н' },
    { n: 'м', c: 'М' }
  ],
};

const keyboardNumberSpecialRu: any = {
  1: [
    { n: '1' },
    { n: '2' },
    { n: '3' },
    { n: '4' },
    { n: '5' },
    { n: '6' },
    { n: '7' },
    { n: '8' },
    { n: '9' },
    { n: '0' },
    { n: ',' },
  ],
  2: [
    { n: '.' },
    { n: ';' },
    { n: ':' },
    { n: '!' },
    { n: '=' },
    { n: '/' },
    { n: '(' },
    { n: ')' },
    { n: '[' },
    { n: ']' },
    { n: '-' }
  ],
  3: [
    { n: '_' },
    { n: '*' },
    { n: '>' },
    { n: '<' },
    { n: '#' },
    { n: '?' },
    { n: '+' },
    { n: '&' },
  ]
};

@Injectable({
  providedIn: 'root'
})
export class LocaleService {
  /*
  * UI texts and methods to update texts to the UI.
  */
  localeTextFi: Array<any> = [
    { id: 'toolbarText', text: 'Kun sisältö ratkaisee' },
    { id: 'nextProgramsText', text: 'Tulossa kanavalta' },
    { id: 'dateElem', text: 'Tänään' },
    { id: 'modalQuestionText', text: 'Sulje Taivas TV7 sovellus' },
    { id: 'exitYesButton', text: 'OK' },
    { id: 'exitCancelButton', text: 'Peruuta' },
    { id: 'comingProgramsText', text: 'Tulossa kanavalta' },
    { id: 'tvIconText', text: 'Netti-TV' },
    { id: 'archiveIconText', text: 'Arkisto' },
    { id: 'searchIconText', text: 'Haku' },
    { id: 'guideIconText', text: 'TV-opas' },
    { id: 'favoritesIconText', text: 'Suosikit' },
    { id: 'channelInfoIconText', text: 'Taivas TV7' },
    { id: 'platformInfoIconText', text: 'Tietoja' },
    { id: 'savedSearchTitle', text: 'Haku' },
    { id: 'searchText', text: 'Hae' },
    { id: 'searchResultText', text: 'Haku tulokset' },
    { id: 'clearText', text: 'Tyhjennä' },
    { id: 'recommendedProgramsText', text: 'Suosittelemme' },
    { id: 'mostViewedProgramsText', text: 'Katsotuimmat' },
    { id: 'newestProgramsText', text: 'Uusimmat' },
    { id: 'categoryText', text: 'Kategoria' },
    { id: 'categoriesText', text: 'Kategoriat' },
    { id: 'topicalSeriesText', text: 'Ajankohtaiset sarjat' },
    { id: 'categoryBackText', text: 'Takaisin' },
    { id: 'episodeText', text: 'Jakso' },
    { id: 'seriesText', text: 'Sarja' },
    { id: 'durationText', text: 'Kesto' },
    { id: 'firstBroadcastText', text: 'Ensiesitys' },
    { id: 'favoritesText', text: 'Suosikit' },
    { id: 'noFavoritesText', text: 'Ei lisättyjä suosikkeja' },
    { id: 'noHitsText', text: 'Ei hakuosumia' },
    { id: 'addedToFavoritesText', text: 'Lisätty suosikkeihin' },
    { id: 'removedFromFavoritesText', text: 'Poistettu suosikeista' },
    { id: 'aspectRatioText', text: 'Kuvasuhde' },
    { id: 'copyrightText', text: 'Copyright © Taivas TV7. Unauthorized publication of programs or subtitles is prohibited.' },
    { id: 'deleteConfigurationsText', text: 'Delete configurations' },
    { id: 'confirmationQuestionText', text: 'Delete configuration?' },
    { id: 'videoStatusConfiguationText', text: 'Video status' },
    { id: 'favoritesConfigurationText', text: 'Favorites' },
    { id: 'searchHistoryConfigurationText', text: 'Search history' },
    { id: 'confirmationOkButton', text: 'OK' },
    { id: 'confirmationCancelButton', text: 'Cancel' },
    { id: 'itemsText', text: 'items' }
  ];

  localeTextEt: Array<any> = [
    { id: 'toolbarText', text: 'Kui sisu on oluline' },
    { id: 'nextProgramsText', text: 'Tulekul' },
    { id: 'dateElem', text: 'Täna' },
    { id: 'modalQuestionText', text: 'Sulge TV7 äpp' },
    { id: 'exitYesButton', text: 'OK' },
    { id: 'exitCancelButton', text: 'Tühista' },
    { id: 'comingProgramsText', text: 'Tulekul' },
    { id: 'tvIconText', text: 'Neti-TV' },
    { id: 'archiveIconText', text: 'Arhiiv' },
    { id: 'searchIconText', text: 'Otsing' },
    { id: 'guideIconText', text: 'Saatekava' },
    { id: 'favoritesIconText', text: 'Lemmikud' },
    { id: 'channelInfoIconText', text: 'Taevas TV7' },
    { id: 'platformInfoIconText', text: 'Infot' },
    { id: 'savedSearchTitle', text: 'Otsing' },
    { id: 'searchText', text: 'Otsi' },
    { id: 'searchResultText', text: 'Otsingu tulemused' },
    { id: 'clearText', text: 'Kustuta' },
    { id: 'recommendedProgramsText', text: 'Vaata arhiivist' },
    { id: 'mostViewedProgramsText', text: 'Vaadatuimad' },
    { id: 'newestProgramsText', text: 'Kõige uuemad' },
    { id: 'categoryText', text: 'Kategooria' },
    { id: 'categoriesText', text: 'Kategooriad' },
    { id: 'topicalSeriesText', text: 'Praegused sarjad' },
    { id: 'categoryBackText', text: 'Tagasi' },
    { id: 'episodeText', text: 'Osa' },
    { id: 'seriesText', text: 'Saatesari' },
    { id: 'durationText', text: 'Kestus' },
    { id: 'firstBroadcastText', text: 'Esimene saade' },
    { id: 'favoritesText', text: 'Lemmikud' },
    { id: 'noFavoritesText', text: 'Lemmikuid pole lisatud' },
    { id: 'noHitsText', text: 'Otsingutulemusi pole' },
    { id: 'addedToFavoritesText', text: 'Lisatud lemmikutesse' },
    { id: 'removedFromFavoritesText', text: 'Lemmikutest eemaldatud' },
    { id: 'aspectRatioText', text: 'Kuvasuhe' },
    { id: 'copyrightText', text: 'Copyright © Taevas TV7. Unauthorized publication of programs or subtitles is prohibited.' },
    { id: 'deleteConfigurationsText', text: 'Delete configurations' },
    { id: 'confirmationQuestionText', text: 'Delete configuration?' },
    { id: 'videoStatusConfiguationText', text: 'Video status' },
    { id: 'favoritesConfigurationText', text: 'Favorites' },
    { id: 'searchHistoryConfigurationText', text: 'Search history' },
    { id: 'confirmationOkButton', text: 'OK' },
    { id: 'confirmationCancelButton', text: 'Cancel' },
    { id: 'itemsText', text: 'items' }
  ];

  localeTextRu: Array<any> = [
    { id: 'toolbarText', text: 'Когда содержание имеет значение' },
    { id: 'nextProgramsText', text: 'Скоро на телеканале' },
    { id: 'dateElem', text: 'Сегодня' },
    { id: 'modalQuestionText', text: 'Закрыть приложение ТВ7' },
    { id: 'exitYesButton', text: 'OK' },
    { id: 'exitCancelButton', text: 'Отмена' },
    { id: 'comingProgramsText', text: 'Скоро на телеканале' },
    { id: 'tvIconText', text: 'Веб-ТВ' },
    { id: 'archiveIconText', text: 'Видеоархив' },
    { id: 'searchIconText', text: 'Поиск' },
    { id: 'guideIconText', text: 'Телепрограмму' },
    { id: 'favoritesIconText', text: 'Избранные' },
    { id: 'channelInfoIconText', text: 'Небеса ТВ7' },
    { id: 'platformInfoIconText', text: 'Информация' },
    { id: 'savedSearchTitle', text: 'Поиск' },
    { id: 'searchText', text: 'Найти' },
    { id: 'searchResultText', text: 'Результаты поиска' },
    { id: 'clearText', text: 'стере́ть' },
    { id: 'recommendedProgramsText', text: 'Рекомендуемые' },
    { id: 'mostViewedProgramsText', text: 'Самые популярные' },
    { id: 'newestProgramsText', text: 'Новые' },
    { id: 'categoryText', text: 'Категория' },
    { id: 'categoriesText', text: 'Категории' },
    { id: 'topicalSeriesText', text: 'Текущая серия' },
    { id: 'categoryBackText', text: 'Обратно' },
    { id: 'episodeText', text: 'Выпуск' },
    { id: 'seriesText', text: 'Серия' },
    { id: 'durationText', text: 'Продолжительность' },
    { id: 'firstBroadcastText', text: 'Первая трансляция' },
    { id: 'favoritesText', text: 'Избранные' },
    { id: 'noFavoritesText', text: 'Список Избранных пуст' },
    { id: 'noHitsText', text: 'Ничего не найдено' },
    { id: 'addedToFavoritesText', text: 'Добавлено в Избранные' },
    { id: 'removedFromFavoritesText', text: 'Удалено из Избранных' },
    { id: 'aspectRatioText', text: 'Соотношение сторон' },
    { id: 'copyrightText', text: 'Copyright © Небеса ТВ7. Unauthorized publication of programs or subtitles is prohibited.' },
    { id: 'deleteConfigurationsText', text: 'Delete configurations' },
    { id: 'confirmationQuestionText', text: 'Delete configuration?' },
    { id: 'videoStatusConfiguationText', text: 'Video status' },
    { id: 'favoritesConfigurationText', text: 'Favorites' },
    { id: 'searchHistoryConfigurationText', text: 'Search history' },
    { id: 'confirmationOkButton', text: 'OK' },
    { id: 'confirmationCancelButton', text: 'Cancel' },
    { id: 'itemsText', text: 'items' }
  ];

  localeTextSv: Array<any> = [
    { id: 'toolbarText', text: 'När innehållet avgör' },
    { id: 'nextProgramsText', text: 'Kommande program på kanalen' },
    { id: 'dateElem', text: 'I dag' },
    { id: 'modalQuestionText', text: 'Stäng Himlen TV7-appen' },
    { id: 'exitYesButton', text: 'OK' },
    { id: 'exitCancelButton', text: 'Avbryt' },
    { id: 'comingProgramsText', text: 'Kommande program på kanalen' },
    { id: 'tvIconText', text: 'TV-kanal' },
    { id: 'archiveIconText', text: 'Play-arkiv' },
    { id: 'searchIconText', text: 'Sök' },
    { id: 'guideIconText', text: 'Tablå' },
    { id: 'favoritesIconText', text: 'Favoriter' },
    { id: 'channelInfoIconText', text: 'Himlen TV7' },
    { id: 'platformInfoIconText', text: 'Information' },
    { id: 'savedSearchTitle', text: 'Sök' },
    { id: 'searchText', text: 'Sök' },
    { id: 'searchResultText', text: 'Sökresultat' },
    { id: 'clearText', text: 'Radera' },
    { id: 'recommendedProgramsText', text: 'Rekommenderat i arkivet' },
    { id: 'mostViewedProgramsText', text: 'Populärast just nu' },
    { id: 'newestProgramsText', text: 'Senaste' },
    { id: 'categoryText', text: 'Kategori' },
    { id: 'categoriesText', text: 'Kategorier' },
    { id: 'topicalSeriesText', text: 'Aktuella serier' },
    { id: 'categoryBackText', text: 'Tillbaka' },
    { id: 'episodeText', text: 'Avsnitt' },
    { id: 'seriesText', text: 'Programserie' },
    { id: 'durationText', text: 'Längd' },
    { id: 'firstBroadcastText', text: 'Första sändningen' },
    { id: 'favoritesText', text: 'Favoriter' },
    { id: 'noFavoritesText', text: 'Inga favoriter valda' },
    { id: 'noHitsText', text: 'Inga sökträffar' },
    { id: 'addedToFavoritesText', text: 'Tillagd bland favoriter' },
    { id: 'removedFromFavoritesText', text: 'Borttagen från favoriter' },
    { id: 'aspectRatioText', text: 'Bildförhållande' },
    { id: 'copyrightText', text: 'Copyright © Himlen TV7. Unauthorized publication of programs or subtitles is prohibited.' },
    { id: 'deleteConfigurationsText', text: 'Delete configurations' },
    { id: 'confirmationQuestionText', text: 'Delete configuration?' },
    { id: 'videoStatusConfiguationText', text: 'Video status' },
    { id: 'favoritesConfigurationText', text: 'Favorites' },
    { id: 'searchHistoryConfigurationText', text: 'Search history' },
    { id: 'confirmationOkButton', text: 'OK' },
    { id: 'confirmationCancelButton', text: 'Cancel' },
    { id: 'itemsText', text: 'items' }
  ];

  constructor(private commonService: CommonService) { }

  getLocaleTextById(id: string): string {
    const localeData: any = this.getLocaleData();

    const result = localeData.find(item => {
      return item.id === id
    });

    if (result) {
      return result.text;
    }
    else {
      return 'Text not found';
    }
  }

  getElementById(id: string): any {
    return this.commonService.getElementById(id);
  }

  setLocaleText(id: string): void {
    if (!id) {
      return;
    }

    const localeText = this.getLocaleTextById(id);
    if (!localeText) {
      return;
    }

    const elem = this.getElementById(id);
    if (!elem) {
      return;
    }

    elem.innerHTML = localeText;
  }

  getSelectedLocale(): string {
    return selectedLocale;
  }

  getArchiveLanguage(): string {
    let lang: string = '';
    if (selectedLocale === localeFi) {
      lang = archiveLanguageFi;
    }
    else if (selectedLocale === localeEt) {
      lang = archiveLanguageEt;
    }
    else if (selectedLocale === localeRu) {
      lang = archiveLanguageRu;
    }
    else if (selectedLocale === localeSv) {
      lang = archiveLanguageSv;
    }

    return lang;
  }

  getArchiveUrl(): string {
    let url: string = '';
    if (selectedLocale === localeFi) {
      url = archiveUrlFi;
    }
    else if (selectedLocale === localeEt) {
      url = archiveUrlEt;
    }
    else if (selectedLocale === localeRu) {
      url = archiveUrlRu;
    }
    else if (selectedLocale === localeSv) {
      url = archiveUrlSv;
    }

    return url;
  }

  getChannelUrl(): string {
    let url: string = '';
    if (selectedLocale == localeFi) {
      url = channelUrlFi;
    }
    else if (selectedLocale == localeEt) {
      url = channelUrlEt;
    }
    else if (selectedLocale == localeRu) {
      url = channelUrlRu;
    }
    else if (selectedLocale == localeSv) {
      url = channelUrlSv;
    }

    return url;
  }

  getArchivePageImage(): string {
    let logo: string = '';
    if (selectedLocale === localeFi) {
      logo = archivePageImageTaivas;
    }
    else if (selectedLocale === localeEt) {
      logo = archivePageImageTaevas;
    }
    else if (selectedLocale === localeRu) {
      logo = archivePageImageNebesa;
    }
    else if (selectedLocale === localeSv) {
      logo = archivePageImageHimlen;
    }

    return logo;
  }

  getArchivePageCategories(): any {
    let categories: any = '';
    if (selectedLocale === localeFi) {
      categories = archivePageCategoriesFi;
    }
    else if (selectedLocale === localeEt) {
      categories = archivePageCategoriesEt;
    }
    else if (selectedLocale === localeRu) {
      categories = archivePageCategoriesRu;
    }
    else if (selectedLocale === localeSv) {
      categories = archivePageCategoriesSv;
    }
    return categories;
  }

  getChannelLogo(): string {
    let logo: string = '';
    if (selectedLocale === localeFi) {
      logo = logoTaivas;
    }
    else if (selectedLocale === localeEt) {
      logo = logoTaevas;
    }
    else if (selectedLocale === localeRu) {
      logo = logoNebesa;
    }
    else if (selectedLocale === localeSv) {
      logo = logoHimlen;
    }

    return logo;
  }

  getLocaleData(): any {
    let locale: any = null;
    if (selectedLocale === localeFi) {
      locale = this.localeTextFi;
    }
    else if (selectedLocale === localeEt) {
      locale = this.localeTextEt;
    }
    else if (selectedLocale === localeRu) {
      locale = this.localeTextRu;
    }
    else if (selectedLocale === localeSv) {
      locale = this.localeTextSv;
    }

    return locale;
  }

  getKeyboard(): any {
    let keyboard: any = null;
    if (selectedLocale === localeFi || selectedLocale === localeSv) {
      keyboard = { letter: keyboardLetters, special: keyboardNumberSpecial };
    }
    else if (selectedLocale === localeEt) {
      keyboard = { letter: keyboardLettersEt, special: keyboardNumberSpecialEt };
    }
    else if (selectedLocale === localeRu) {
      keyboard = { letter: keyboardLettersRu, special: keyboardNumberSpecialRu };
    }

    return keyboard;
  }

  getAppName(): string {
    return appName;
  }

  getAppVersion(): string {
    return appVersion;
  }
}
